### PR TITLE
fix: when deleting datasets, only delete artifacts that are no longer reference by a dataset.

### PR DESCRIPTION
**Pull request by Bento007**
… reference by a dataset.

## Reason for Change

- https://czi.atlassian.net/browse/VC-4484

## Changes
This pull request implements robust reference-counted deletion of dataset artifacts to ensure that shared artifacts are not accidentally deleted when removing dataset versions. It introduces logic to track which artifacts are referenced by multiple dataset versions and only deletes artifacts that are no longer referenced elsewhere. Several interfaces and method signatures are updated to support passing explicit sets of artifacts to preserve, and new utility methods for artifact reference queries are added.

**Reference-counted artifact deletion and preservation:**

* Added logic in `business.py` to determine which artifacts should be preserved when deleting dataset versions, using a new `_get_artifacts_to_save` method that combines explicit preservation with reference counting based on database queries. This ensures that shared artifacts are not deleted if referenced by other dataset versions.
* Updated the deletion workflow in `publish_collection_version` and related methods to use the new artifact preservation logic, preventing accidental deletion of artifacts still in use. [[1]](diffhunk://#diff-df36ae23a2a7f265486f5569606db86dbd40811ca8b99fda3aafb293d1b1d379R1227-R1243) [[2]](diffhunk://#diff-df36ae23a2a7f265486f5569606db86dbd40811ca8b99fda3aafb293d1b1d379L1090-R1114)
* Modified the artifact deletion methods in both business and persistence layers to accept an explicit set of artifacts to save (`artifacts_to_save`), updating method signatures and call sites accordingly. [[1]](diffhunk://#diff-df36ae23a2a7f265486f5569606db86dbd40811ca8b99fda3aafb293d1b1d379L1040-R1044) [[2]](diffhunk://#diff-df36ae23a2a7f265486f5569606db86dbd40811ca8b99fda3aafb293d1b1d379L1054-R1054) [[3]](diffhunk://#diff-9834c2af69c51183bbc452bf4a8babfe8045df83df6bf0964729b9ce448ccd69L102-R104) [[4]](diffhunk://#diff-9e9450f1ed6950443571ebe240d2236c1f95e732079e0d82bc6fef56fbe1a149L137-R146)

**Persistence layer enhancements:**

* Added `get_artifact_references` to the persistence layer and its interface, allowing queries for which dataset versions reference a given set of artifacts. This is used by the business logic to determine which artifacts are still in use. [[1]](diffhunk://#diff-7f3edce24972d8614da437bd88d9eaaa42d08199b24816370c50f7dc446b6d6bR864-R895) [[2]](diffhunk://#diff-9e9450f1ed6950443571ebe240d2236c1f95e732079e0d82bc6fef56fbe1a149R226-R233)
* Updated `delete_dataset_versions` and related methods in the persistence layer to support skipping deletion of artifacts that are still referenced, and to propagate the `artifacts_to_save` parameter throughout the deletion process. [[1]](diffhunk://#diff-7f3edce24972d8614da437bd88d9eaaa42d08199b24816370c50f7dc446b6d6bL627-R670) [[2]](diffhunk://#diff-9e9450f1ed6950443571ebe240d2236c1f95e732079e0d82bc6fef56fbe1a149L137-R146)

**Entity and typing improvements:**

* Implemented `__hash__` methods for `EntityId` and `DatasetArtifact` to allow them to be used in sets for efficient artifact reference tracking and comparison. [[1]](diffhunk://#diff-bf4c96a258f8e19ff47c3b8374a57b21472f5fc492a5840665ce2cb38b5bf586R140-R142) [[2]](diffhunk://#diff-bf4c96a258f8e19ff47c3b8374a57b21472f5fc492a5840665ce2cb38b5bf586R177-R179)
* Updated type hints and imports throughout the business and persistence layers to use `Set` where appropriate for artifact collections. [[1]](diffhunk://#diff-9834c2af69c51183bbc452bf4a8babfe8045df83df6bf0964729b9ce448ccd69L1-R1) [[2]](diffhunk://#diff-7f3edce24972d8614da437bd88d9eaaa42d08199b24816370c50f7dc446b6d6bL7-R7) [[3]](diffhunk://#diff-766d914dd42c86f642d416c3e803a78eab7c8c8a14ec64f2b5f31d1358f4be70L4-R4) [[4]](diffhunk://#diff-9e9450f1ed6950443571ebe240d2236c1f95e732079e0d82bc6fef56fbe1a149L2-R2)

These changes collectively ensure safe deletion of dataset artifacts, preventing data loss when artifacts are shared between multiple dataset versions.

### CODE REVIEW

**Comment by Bento007** [comment]
@nayib-jose-gloria we do still need a DB call incase there are other datasets outside the collection scope. I did take our advice and consolidate as much of the logic into the business side as possible and added unit tests for those new functions. Here's what changed:

# Summary of Changes: 133afb5 → 850c0f2

## 🎯 Key Architectural Change

The major change is a **refactoring of the artifact retention logic** from persistence layer to business layer, introducing **reference counting** to prevent deleting shared artifacts.

---

## 📊 High-Level Stats

- **7 commits** between 133afb5 and 850c0f2
- **+391 insertions, -84 deletions** across 7 files
- **+174 new test lines** to cover the enhanced logic

### Files Changed

| File | Changes |
|------|---------|
| `backend/layers/business/business.py` | +106/-? |
| `backend/layers/business/business_interface.py` | +6/-? |
| `backend/layers/common/entities.py` | +6/- |
| `backend/layers/persistence/persistence.py` | +84/-? |
| `backend/layers/persistence/persistence_interface.py` | +19/-? |
| `backend/layers/persistence/persistence_mock.py` | +80/-? |
| `tests/unit/backend/layers/business/test_business.py` | +174/-? |

---

## 🔄 Major Refactorings

### 1. Reference Counting Moved from Persistence → Business Layer

**Before (at 133afb5):**
- Persistence layer did reference counting inline during deletion using `Counter`
- Logic: "Only delete artifacts referenced once"
- Located in `_delete_dataset_version_and_artifact_rows()`

**After (at 850c0f2):**
- Business layer explicitly determines which artifacts to save via new method `_get_artifacts_to_save()`
- Persistence layer receives `artifacts_to_save` set and simply excludes them from deletion
- New method `get_artifact_references()` in persistence layer provides pure data query
- Better separation of concerns: business logic decisions vs. data operations

### 2. New Helper Method: `_get_artifacts_to_save()`

A new business layer method that combines:

1. **Explicit saves**: Artifacts shared between kept and deleted versions
2. **Reference counting**: Artifacts referenced by other (non-deleted) dataset versions

```python
def _get_artifacts_to_save(
    self,
    artifacts_being_deleted: List[DatasetArtifact],
    dataset_versions_being_deleted: List[DatasetVersion],
    explicitly_saved_artifacts: Set[DatasetArtifact] = None,
) -> Set[DatasetArtifact]:
    """
    Determines which artifacts must be saved. Combines explicit saves with reference counting.

    :param artifacts_being_deleted: All artifacts from dataset versions being deleted
    :param dataset_versions_being_deleted: Dataset versions being deleted
    :param explicitly_saved_artifacts: Optional set of artifacts explicitly marked to save
    :return: Set of artifacts that must be saved (not deleted from S3 or database)
    """
```

### 3. New Method: `_cleanup_unpublished_versions_after_publish()`

Extracted cleanup logic from `publish_collection_version()` into its own method for:

- Better testability
- Clearer separation of concerns
- More comprehensive documentation

```python
def _cleanup_unpublished_versions_after_publish(
    self,
    version: CollectionVersionWithDatasets,
    date_of_last_publish: datetime,
    is_auto_version: bool,
) -> None:
    """
    Handles cleanup of unpublished dataset versions after collection publication.
    Determines which versions to keep (including auto_version revisions) and
    safely deletes unused versions while preserving shared artifacts.
    """
```

### 4. Enhanced Type Safety

- Changed `**kwargs` to explicit `artifacts_to_save: Set[DatasetArtifact]` parameter
- Added `Set` to imports across multiple files
- Added `__hash__()` methods to `EntityId` and `DatasetArtifact` classes to support set operations

---

## 🔧 API Changes

### Business Layer

```python
# Added explicit parameter (replaced **kwargs)
def delete_dataset_versions(
    self, 
    dataset_versions: List[DatasetVersion], 
    artifacts_to_save: Set[DatasetArtifact] = None  # NEW
) -> None:

def delete_dataset_version_assets(
    self,
    dataset_versions: List[DatasetVersion],
    artifacts_to_save: Set[DatasetArtifact] = None  # NEW
) -> None:
```

### Persistence Layer

```python
# NEW method for reference data
def get_artifact_references(
    self, 
    artifact_ids: List[DatasetArtifactId]
) -> Dict[str, List[str]]:
    """
    Returns which dataset versions reference each artifact. Pure data query with no business logic.
    
    :param artifact_ids: List of artifact IDs to check references for
    :return: Dictionary mapping artifact_id (str) to list of dataset_version_ids (str) that reference it
    """

# Updated signature
def delete_dataset_versions(
    self,
    dataset_versions: List[Union[DatasetVersionId, DatasetVersion]],
    artifacts_to_save: Optional[Set[DatasetArtifact]] = None  # NEW
) -> None:

# REMOVED unused method
def get_all_artifacts(self) -> List[DatasetArtifact]:  # ❌ DELETED
```

---

## ✅ Testing Improvements

Added new test class `TestArtifactCleanupHelpers` with comprehensive tests:

1. **`test_get_artifacts_to_save_empty_list`** - Edge case handling
2. **`test_get_artifacts_to_save_only_explicit_saves`** - Explicit saves work
3. **`test_get_artifacts_to_save_with_reference_counting`** - Reference counting works
4. **`test_get_artifacts_to_save_artifact_only_in_deleted_versions`** - Unique artifacts deleted
5. **`test_get_artifacts_to_save_combines_explicit_and_reference_counting`** - Both work together
6. **`test_cleanup_unpublished_versions_after_publish_basic`** - Basic cleanup test
7. **`test_cleanup_unpublished_versions_preserves_shared_artifacts`** - Shared artifacts preserved
8. **`test_cleanup_with_auto_version_and_open_revision`** - Auto-version edge case

---

## 🔍 Code Quality Improvements

1. **Naming consistency**: `artifact_to_save` → `artifacts_to_save` (plural, consistent)
3. **Removed dead code**: `get_all_artifacts()` method was unused
4. **Type hints**: More explicit type annotations (`Set`, `Dict`, etc.)

File: backend/layers/business/business.py [hunk 1]
```diff
@@ -1225,17 +1252,61 @@ def publish_collection_version(
...
+        :param artifacts_being_deleted: All artifacts from dataset versions being deleted
+        :param dataset_versions_being_deleted: Dataset versions being deleted
+        :param explicitly_saved_artifacts: Optional set of artifacts explicitly marked to save (e.g., shared with kept versions)
+        :return: Set of artifacts that must be saved (not deleted from S3 or database)
+        """
+        if not artifacts_being_deleted:
+            return set()
+
+        # Start with explicitly saved artifacts
+        artifacts_to_save = set(explicitly_saved_artifacts or [])
```
Comment (kliu-czi) [inline backend/layers/business/business.py]: Why do we try to convert a default empty list to a set? Can we just declare the empty set instead?

Ex:
`artifacts_to_save = explicitly_saved_artifacts or set()`

File: backend/layers/persistence/persistence.py [hunk 2]
```diff
@@ -624,38 +624,50 @@ def delete_datasets(self, datasets: List[Union[DatasetId, CanonicalDataset]]) ->
...
-        )
-        session.execute(dataset_version_delete_statement)
+        dataset_version_ids = [str(dv_row.id) for dv_row in dataset_version_rows]
+        all_artifact_ids = {str(artifact) for dv_row in dataset_version_rows for artifact in dv_row.artifacts}
+
+        # Exclude artifacts_to_save
+        artifacts_to_save_ids = {str(a.id) for a in (artifacts_to_save or [])}
+        artifact_ids_to_delete = [aid for aid in all_artifact_ids if aid not in artifacts_to_save_ids]
+
+        session.execute(delete(DatasetArtifactTable).where(DatasetArtifactTable.id.in_(artifact_ids_to_delete)))
```
[the remark this answers is not in the document]
Comment (Bento007) [reply backend/layers/persistence/persistence.py]: Yes, since they happen within the same SQLAlchemy session.

File: backend/layers/business/business.py [hunk 3]
```diff
@@ -1225,17 +1252,61 @@ def publish_collection_version(
...
+        :param artifacts_being_deleted: All artifacts from dataset versions being deleted
+        :param dataset_versions_being_deleted: Dataset versions being deleted
+        :param explicitly_saved_artifacts: Optional set of artifacts explicitly marked to save (e.g., shared with kept versions)
+        :return: Set of artifacts that must be saved (not deleted from S3 or database)
+        """
+        if not artifacts_being_deleted:
+            return set()
+
+        # Start with explicitly saved artifacts
+        artifacts_to_save = set(explicitly_saved_artifacts) or set()
```
Comment (Bento007) [inline backend/layers/business/business.py]:
```suggestion
        artifacts_to_save = set(explicitly_saved_artifacts) if explicitly_saved_artifacts else set()
```
